### PR TITLE
Add purescript maker

### PR DESCRIPTION
**Pull request by vaibhavsagar**
Instead of #359 this is a simpler PureScript maker based on the Elm maker. This is pretty barebones at the moment but I'd love any and all feedback!

### CODE REVIEW

File: autoload/neomake/makers/ft/purescript.vim [hunk 1]
```diff
@@ -0,0 +1,59 @@
+" vim: ts=4 sw=4 et
+
+function! neomake#makers#ft#purescript#EnabledMakers()
+  return ['pulp']
+endfunction
+
+function! neomake#makers#ft#purescript#pulp()
+    return {
+        \ 'exe': 'pulp',
+        \ 'args': ['build', '--no-psa', '--', '--json-errors'],
```
Comment (blueyed) [inline autoload/neomake/makers/ft/purescript.vim]: `exe` defaults to `pulp`.
Is `--` required in `args` really?  (typically it's used to separate options from filenames, so `--json-errors` would end up being considered a filename?!)

Comment (vaibhavsagar) [inline autoload/neomake/makers/ft/purescript.vim]: In this case it's used to separate options to `pulp` from options to `purs compile` which it uses under the hood.

Comment (blueyed) [inline autoload/neomake/makers/ft/purescript.vim]: I see.  Should have a comment then.

Comment (vaibhavsagar) [inline autoload/neomake/makers/ft/purescript.vim]: You're right, I've added a comment and a link to the issue where I got this information.

**Comment by blueyed** [comment]
LGTM, should have some test though.
See `tests/ft_rust.vader`.

**Comment by vaibhavsagar** [comment]
Noted, I will look into adapting an existing test.

**Comment by blueyed** [comment]
@vaibhavsagar 
ping

**Comment by vaibhavsagar** [comment]
Hey, thanks for the ping! I got busy with life stuff but will look at this over the weekend 😄

**Comment by vaibhavsagar** [comment]
Still a work in progress!

File: tests/ft_purescript.vader [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    \ ]
+
+  let result = neomake#makers#ft#purescript#pulp().process_output({'output': output})
+
+  AssertEqual [], result
+
+Execute (pulp parses error message):
+  new
+  file src/Main.purs
+  file src/Other.purs
```
Comment (blueyed) [inline tests/ft_purescript.vader]: Don't call `file` twice on the same buffer, but should call `new` inbetween maybe?!

**Comment by blueyed** [comment]
Pushed a fixup to fix the indent issue.

What do you plan to do additionally?

**Comment by vaibhavsagar** [comment]
I also need to test the warnings.

File: tests/ft_purescript.vader (hunk 2 above)
Comment (vaibhavsagar) [inline tests/ft_purescript.vader]: Thanks. This is my first ever Vader test, can you tell :smile:

**Comment by blueyed** [comment]
I will look into the bufnr stuff. It seems to be the same as with https://github.com/neomake/neomake/pull/1218.

**Comment by blueyed** [comment]
https://github.com/neomake/neomake/pull/1257 should address this.

btw: I think there is a conceptual error in using a file_mode maker, when in fact it should be a project-based one: file_mode makers use location lists, while otherwise a quickfix list is used.  I am not really sure about how to address this, and think the whole idea of `file_mode` needs to be revisited.

**Comment by vaibhavsagar** [comment]
Thanks for looking into this! I agree that having this as a project-based maker would make more sense. I will look into it.

**Comment by blueyed** [comment]
It is more a general, conceptual issue I think.
E.g. `cargo` was switched to a file_mode maker recently, but I am not sure what the reason was.

**Comment by blueyed** [comment]
Pushed another commit to use the project root in `cwd`, since the output is relative to it.

Needs more fixed to https://github.com/neomake/neomake/pull/1257 though really.

**Comment by vaibhavsagar** [comment]
You've put much more work into this than I have :smile:. I've added a test for warnings.

**Comment by blueyed** [comment]
Merged master, which contains #1257.

Is this OK from your side?
Do you want to make it a project maker already?
My idea about that would be to allow for makers to live in `ft/`, but allow to specify if they are project makers (or rather if they want to use the quickfix or location list).

**Comment by vaibhavsagar** [comment]
I'm happy with the file-based maker for now, I just wanted something simple to start off with and I'm happy for someone else (or future me) to change it to be project-based further down the line.

**Comment by blueyed** [comment]
Thanks!

**Comment by blueyed** [comment]
Please add it to the wiki's list of makers.

**Comment by vaibhavsagar** [comment]
Done, thanks!

**Pull request merged**